cpp/src/tracing.cpp

```diff
@@ -0,0 +1,240 @@
+    options.kind          = opentelemetry::trace::SpanKind::kConsumer;
+
+    // extract context from AMQP message annotations
+    const AMQPTextMapCarrier<proton::message::annotation_map> carrier(
+        &message.message_annotations());
+
+    nostd::shared_ptr<
+        opentelemetry::v0::context::propagation::TextMapPropagator>
+        prop = opentelemetry::context::propagation::GlobalTextMapPropagator::
+            GetGlobalPropagator();
+    context::Context current_ctx =
+        opentelemetry::context::RuntimeContext::GetCurrent();
+
+    context::Context new_context = prop->Extract(carrier, current_ctx);
+    options.parent =
+        opentelemetry::trace::propagation::GetSpan(new_context)->GetContext();
+
+    binary tag_in_binary = d.tag();
+    std::string tag_in_string = std::string(d.tag());
+    std::stringstream ss;
+    for(int i=0; i<(int)tag_in_string.length(); ++i)
+        ss << std::hex << (int)tag_in_binary[i];
+    std::string delivery_tag = ss.str();
+
+    receiver r = d.receiver();
+    source s = r.source();
+    std::string s_addr = s.address();
+
+    transfer tt(d);
+    std::string delivery_state = to_string(tt.state());
+
+    nostd::shared_ptr<trace::Span> span =
+        get_tracer()->StartSpan("amqp-message-received",
+                                {{"Delivery_tag", delivery_tag},
+                                 {"Source_address", s_addr}},
+                                options);
+
+    trace::Scope scope = get_tracer()->WithActiveSpan(span);
+
+    if (delivery_state == "unknown")
+    {
+        span->AddEvent("Delivery state is unknown");
+    }
+    if (delivery_state == "Received")
+    {
+        span->AddEvent("Delivery is received");
+    }
+
+    tag_span[tag_in_binary] = span;
+}
+
+void on_message_end_span(delivery &d) {
+    binary tag = d.tag();
+    nostd::shared_ptr<trace::Span> span = tag_span[tag];
+
+    span->End();
+}
+
+void on_settled_span(delivery &d) {
+    binary tag = d.tag();
+    nostd::shared_ptr<trace::Span> span = tag_span[tag];
+
+    span->End();
+}
+
+void send_span(message &message, binary &tag, tracker &track) {
+
+    opentelemetry::trace::StartSpanOptions options;
+    options.kind = opentelemetry::trace::SpanKind::kProducer;
+
+    binary tag_in_binary = tag;
+    std::string tag_in_string = std::string(tag);
+    std::stringstream ss;
+    for (int i = 0; i < (int)tag_in_string.length(); ++i)
+        ss << std::hex << (int)tag_in_binary[i];
+    std::string delivery_tag = ss.str();
+
+    sender s = track.sender();
+    target t = s.target();
+    std::string t_addr = t.address();
+
+    transfer tt(track);
+    std::string delivery_state = to_string(tt.state());
+
+    nostd::shared_ptr<trace::Span> span = get_tracer()->StartSpan(
+        "amqp-delivery-send",
+        {{"Delivery_tag", delivery_tag}, {"Destination_address", t_addr}},
+        options);
+
+    trace::Scope scope = proton::get_tracer()->WithActiveSpan(span);
```

Comment:
I dont think the send span should be set as the 'active' span, as we dont want it to affect other things (e.g replacing an on_message active span), and/or the life of the scope then seems like it would be in play. We just want to create and retain a span and close it once the message is accepted/other later.

Reply:
As discussed on IRC, it seems that in the opentelemetry-cpp case, the returned 'scope' being destroyed will reinstate the previously active span as the active span, so as that destruction seems to happen before send returns then it will essentially behave as I wanted in terms of not leaving the new span as the 'active'.

Setting the span as active in this case seems largely redundant since we have no other related spans for it to impact and simply inject the context and send the messaage, but if thats required then so be it. It was the knock on effect on the 'active' I was concerned with.